<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Counters for read datasets · MerCounting.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">MerCounting.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../installation/">Installation</a></li><li><a class="tocitem" href="../basic_counting/">Basic types and counting</a></li><li class="is-active"><a class="tocitem" href>Counters for read datasets</a><ul class="internal"><li><a class="tocitem" href="#The-serial_mem-counter-1"><span>The <code>serial_mem</code> counter</span></a></li></ul></li></ul></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../../api/MerCount/">MerCount</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Counters for read datasets</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Counters for read datasets</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/MerCounting.jl/blob/master/docs/src/man/counting_reads.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Counting-kmers-in-read-datasets-1"><a class="docs-heading-anchor" href="#Counting-kmers-in-read-datasets-1">Counting kmers in read datasets</a><a class="docs-heading-anchor-permalink" href="#Counting-kmers-in-read-datasets-1" title="Permalink"></a></h1><p>MerCounting provides some more dedicated counting algorithms for counting kmers in read datasets, particularly ReadDatastores.</p><p>This is because as conceptually simple as counting kmers is, to do it quickly for large sequencing read datasets output by sequencing machines can be difficult for some datasets.</p><p>There are many ways you can try to optimise the kmer counting process, and many kmer counting tools already exist.</p><p>MerCounting provides a <code>Counters</code> submodule, which contains nessecery methods and types required to implement various kinds of kmer counter, as well as exporting a selection of &quot;off-the-shelf&quot; methods that use different counting strategies, one or several of which hopefully suit the user&#39;s dataset and computational resources available. We describe and showcase these below.</p><h2 id="The-serial_mem-counter-1"><a class="docs-heading-anchor" href="#The-serial_mem-counter-1">The <code>serial_mem</code> counter</a><a class="docs-heading-anchor-permalink" href="#The-serial_mem-counter-1" title="Permalink"></a></h2><p>The <a href="man/@ref"><code>serial_mem</code></a> counter is the simplest kmer counter for reads that is provided.</p><p>It counts kmers in each read serially, and stores the kmers and counts entirely in RAM.</p><article class="docstring"><header><a class="docstring-binding" id="MerCounting.Counters.serial_mem" href="#MerCounting.Counters.serial_mem"><code>MerCounting.Counters.serial_mem</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">serial_mem(::Type{M}, input::DatastoreBuffer{&lt;:ReadDatastore}, count_mode::CountMode, range::AbstractRange = 1:length(input)) where {M&lt;:AbstractMer}</code></pre><p>MerCounting&#39;s simplest kmer counting method.</p><p>Build a sorted list (vector) of kmer counts (MerCount), serially and entirely in memory.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This function is a serial and in memory <code>MerCount</code> list builder that can build a kmer count from a ReadsDatastore on its own (if you have memory and time), but it is also intended to be composed into other multi-process or multi-threaded kmer counting strategies.</p><p>This method estimates roughly how many kmers will be generated by the reads specified by <code>range</code> in the dataset. It then pre-allocates an array to contain them. It then collects the kmers, sorts, them, and then collapses them into a list of counts sorted by the kmer.</p><p>So if you want to count kmers and have the resources to throw at it, this is the simplest method, and possibly even the quickest given that simplicity.</p></div></div></div></section><section><div><pre><code class="language-none">serial_mem(::Type{M}, input::ReadDatastore, count_mode::CountMode, range::AbstractRange = 1:length(input)) where {M&lt;:AbstractMer}</code></pre><p>MerCounting&#39;s simplest kmer counting method.</p><p>Build a sorted list (vector) of kmer counts (MerCount), serially and entirely in memory.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This function is a serial and in memory <code>MerCount</code> list builder that can build a kmer count from a ReadsDatastore on its own (if you have memory and time), but it is also intended to be composed into other multi-process or multi-threaded kmer counting strategies.</p><p>This method estimates roughly how many kmers will be generated by the reads specified by <code>range</code> in the dataset. It then pre-allocates an array to contain them. It then collects the kmers, sorts, them, and then collapses them into a list of counts sorted by the kmer.</p><p>So if you want to count kmers and have the resources to throw at it, this is the simplest method, and possibly even the quickest given that simplicity.</p></div></div></div></section></article><p>It is fairly simple to use. First you need a ReadDatastore, if you need to recall how to create one then head over to <a href="https://biojulia.net/ReadDatastores.jl/stable/build-datastores/">here</a> in ReadDatastores.jl&#39;s documentation.</p><p>The example below opens a datastore before using <a href="man/@ref"><code>serial_mem</code></a> to count the kmers in the read datastore.</p><pre><code class="language-julia-repl">julia&gt; using MerCounting

julia&gt; using ReadDatastores

julia&gt; ds = @openreads &quot;ecoli-test-paired.prseq&quot;
ERROR: LoadError: SystemError: opening file &quot;ecoli-test-paired.prseq&quot;: No such file or directory
in expression starting at none:1

julia&gt; kl = Counters.serial_mem(DNAMer{31}, ds, CANONICAL)
ERROR: UndefVarError: ds not defined</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../basic_counting/">« Basic types and counting</a><a class="docs-footer-nextpage" href="../../api/MerCount/">MerCount »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 1 May 2020 15:17">Friday 1 May 2020</span>. Using Julia version 1.3.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
