<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Analysing Kmer count data · KmerAnalysis.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">KmerAnalysis.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../installation/">Installation</a></li><li><a class="tocitem" href="../basic_counting/">Basic types and counting</a></li><li><a class="tocitem" href="../counting_reads/">Counters for read datasets</a></li><li class="is-active"><a class="tocitem" href>Analysing Kmer count data</a><ul class="internal"><li><a class="tocitem" href="#K-mer-frequency-spectra-1"><span>K-mer frequency spectra</span></a></li><li><a class="tocitem" href="#Kmer-count-projection-1"><span>Kmer count projection</span></a></li></ul></li></ul></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../../api/MerCount/">MerCount</a></li><li><a class="tocitem" href="../../api/KmerFrequencySpectra/">KmerFrequencySpectra</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Analysing Kmer count data</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Analysing Kmer count data</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/KmerAnalysis.jl/blob/master/docs/src/man/analyses.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Analysing-k-mer-count-data-1"><a class="docs-heading-anchor" href="#Analysing-k-mer-count-data-1">Analysing k-mer count data</a><a class="docs-heading-anchor-permalink" href="#Analysing-k-mer-count-data-1" title="Permalink"></a></h1><p>Once you have k-mer count data, there are a number of useful analyses that may be done. Let&#39;s go through a few together.</p><h2 id="K-mer-frequency-spectra-1"><a class="docs-heading-anchor" href="#K-mer-frequency-spectra-1">K-mer frequency spectra</a><a class="docs-heading-anchor-permalink" href="#K-mer-frequency-spectra-1" title="Permalink"></a></h2><p>A k-mer frequency spectra is a representation of sequencing data showing how many k-mers appear a certain number of times. In other words, how many k-mers in the dataset have a count of 1, how many have a count of 10, and so on.</p><p>When plotted, the frequency of occurence is plotted on the x-axis, and the number of k-mers on the y-axis.</p><p>The k-mer spectra is composed of distributions representing groups of motifs at different frequencies in the sample, plus biases. Given not too many biases, the shape of the distributions provides a useful set of properties describing the biological sample, the sequencing process and the amount of useful data in the dataset.</p><p>Let&#39;s take a look at a simple spectra. I&#39;m going to use a small artificial dataset of a bacterial-like genome I generated using <a href="https://bioinfologics.github.io/Pseudoseq.jl/stable/">Pseudoseq.jl</a>.</p><p>I first created a random single chromosome genome, and then simulated a paired-end sequencing experiment to generate the kind of raw data we would get from a sequencing experiment.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>I&#39;m doing this with a small artificial genome to be kind to the CI infrastructure that builds these docs!!!</p><p>But you could follow along and try yourself with paired-end read data from a sequencing experiment from e.coli.</p></div></div><p>Here I&#39;m going to load in these reads, count the k-mers with the <a href="man/@ref"><code>SerialMemCounter</code></a> k-mer counter as the dataset is not that large, and then compute the 1D frequency spectra.</p><pre><code class="language-julia-repl">julia&gt; using KmerAnalysis, ReadDatastores, BioSequences

julia&gt; reads = @openreads &quot;../../../test/fakemicrobe.prseq&quot;
Paired Read Datastore &#39;fakemicrobe-pe&#39;: 83850 reads (41925 pairs)

julia&gt; counter = serial_mem(DNAMer{31}, CANONICAL)
(::KmerAnalysis.SerialMemCounter{BioSequences.Mer{BioSequences.DNAAlphabet{2},31},Canonical}) (generic function with 1 method)

julia&gt; kmer_counts = counter(reads)
1561193-element Array{MerCount{BioSequences.Mer{BioSequences.DNAAlphabet{2},31}},1}:
 AAAAAAAAAAAGGTATCATCCTGTAGCTAGT occurs 1 times
 AAAAAAAAAAGGTATCATCCTGTAGCTAGTG occurs 1 times
 AAAAAAAAAATCTCGGACAGCGCCGGTCAGC occurs 15 times
 AAAAAAAAAATCTCTGACAGCGCCGGTCAGC occurs 1 times
 AAAAAAAAAATGCGACGTGTCCCAGCAGAGC occurs 16 times
 AAAAAAAAAATGCGACGTGTCCGAGCAGAGC occurs 1 times
 AAAAAAAAACGCCTAGTAGTACTTTAATGCG occurs 17 times
 AAAAAAAAACGCTGGTATTCGTGTTTGGTGG occurs 10 times
 AAAAAAAAAGACCAAGAGACATCGTGCTGGG occurs 23 times
 AAAAAAAAAGGTATCATCCTGTAGCTAGTGT occurs 1 times
 ⋮
 TTTTAGGAGAAAAGCGCTTGCGCAAGCAAAA occurs 18 times
 TTTTCACACAGTCTAAGATTATTCAAGAAAA occurs 25 times
 TTTTCAGCAGATCCGCCGATCTCGCTCAAAA occurs 1 times
 TTTTCAGCAGATCCGCCGATCTCGCTGAAAA occurs 22 times
 TTTTCGTACACCGGGTAAGTAACGTTAAAAA occurs 1 times
 TTTTCGTACTCCGGGTAAGTAACGTTAAAAA occurs 16 times
 TTTTGACTCTCACTACCTAGGCATCTAAAAA occurs 17 times
 TTTTGAGGTGCAAGCCGGAACAAAAACAAAA occurs 22 times
 TTTTGCTGATCACGTGAGACGATGCACAAAA occurs 20 times

julia&gt; my_spectra = spectra(kmer_counts)
ERROR: MethodError: no method matching isless(::typeof(min), ::UInt8)
Closest candidates are:
  isless(!Matched::Missing, ::Any) at missing.jl:87
  isless(!Matched::AbstractFloat, ::Real) at operators.jl:158
  isless(!Matched::Real, ::Real) at operators.jl:346
  ...</code></pre><p>I could have also created the counter and passed it to the <a href="../../api/KmerFrequencySpectra/#KmerAnalysis.spectra"><code>spectra</code></a> method along with the reads like so:</p><pre><code class="language-julia-repl">julia&gt; my_spectra = spectra(counter, reads)
ERROR: MethodError: no method matching isless(::typeof(min), ::UInt8)
Closest candidates are:
  isless(!Matched::Missing, ::Any) at missing.jl:87
  isless(!Matched::AbstractFloat, ::Real) at operators.jl:158
  isless(!Matched::Real, ::Real) at operators.jl:346
  ...</code></pre><h2 id="Kmer-count-projection-1"><a class="docs-heading-anchor" href="#Kmer-count-projection-1">Kmer count projection</a><a class="docs-heading-anchor-permalink" href="#Kmer-count-projection-1" title="Permalink"></a></h2><p>KmerAnalysis includes a k-mer count container type called IndexedCounts. Indexed counts are different to a vector or other container of <code>MerCount</code>, in that IndexedCounts store k-mer counts from many sources (often different read datasets) indexed against a reference dataset (usually a genome graph or set of contigs).</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../counting_reads/">« Counters for read datasets</a><a class="docs-footer-nextpage" href="../../api/MerCount/">MerCount »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Monday 25 May 2020 01:39">Monday 25 May 2020</span>. Using Julia version 1.3.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
